src/buffer-pool.hpp

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <array>
+#include <mutex>
+#include <atomic>
+
+class buffer_pool_buf
+{
+public:
+    bool ready_capture() const
+    {
+        return released;
+    }
+
+    bool ready_encode() const
+    {
+        return available;
+    }
+
+    std::atomic<bool> released{true}; // if the buffer can be used to store new pending frames
+    std::atomic<bool> available{false}; // if the buffer can be used to feed the encoder
+};
+
+template <class T, int N>
+class buffer_pool
+{
+public:
+    size_t size() const
+    {
+        return N;
+    }
+
+    const T& at(size_t i) const
+    {
+        return bufs[i];
+    }
+
+    T& capture()
+    {
+        return bufs[capture_idx];
+    }
+
+    T& encode()
+    {
+        return bufs[encode_idx];
+    }
+
+    T& next_capture()
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+        bufs[capture_idx].released = false;
+        bufs[capture_idx].available = true;
+        capture_idx = (capture_idx + 1) % bufs_size;
+        if (capture_idx == 0 && !bufs[0].ready_capture() && bufs_size < N)
```

Comment:
Since we basically have a circular buffer, I wonder why we need to enlarge only when `capture_idx==0`. Is it not possible that the buffer is full at some other point in time? For example the following scenario (with say 2 buffers):

- Capture frame 0
- Capture frame 1
- Encode frame 0
- Capture frame 0 -> capture_idx is incremented to 1, but it is not yet ready for capture -> what do we do?